File: sexp_formatter_minimal.py
#!/usr/bin/env python3
# KiCADv8 Style Prettify S-Expression Formatter (sexp formatter) (minimal logic version)
# By Brian Khuu, 2024
# This script reformats KiCad-like S-expressions to match a specific formatting style.
# Note: This script modifies formatting only; it does not perform linting or validation.

import argparse
from pathlib import Path

def prettify_sexpr_minimal(sexpr_str):
    """
    Prettifies KiCad-like S-expressions according to a KiCADv8-style formatting (minimal logic version)

    Args:
        sexpr_str (str): The input S-expression string to be formatted.

    Returns:
        str: The prettified S-expression string.

    Example:
        # Input S-expression string
        sexpr = "(module (fp_text \"example\"))"

        # Settings for compact element handling
        compact_settings = [{"prefix": "pts", "elements per line": 4}]

        # Prettify the S-expression
        formatted_sexpr = prettify_sexpr_minimal(sexpr, compact_settings)
        print(formatted_sexpr)
    """

    indent = 0
    result = []

    in_quote = False
    escape_next_char = False
    singular_element = False

    # Iterate through the s-expression and apply formatting
    for char in sexpr_str:

        if char == '"' or in_quote:
            # Handle quoted strings, preserving their internal formatting
            result.append(char)
            if escape_next_char:
                escape_next_char = False
            elif char == '\\':
                escape_next_char = True
            elif char == '"':
                in_quote = not in_quote

        elif char == '(':
            # Check for compact element handling
            result.append('\n' + '\t' * indent + '(')
            singular_element = True
            indent += 1

        elif char == ')':
            # Handle closing elements
            indent -= 1
            if singular_element:
                result.append(')')
                singular_element = False
            else:
                result.append('\n' + '\t' * indent + ')')

        elif char.isspace():
            # Handling spaces
            if result and not result[-1].isspace() and result[-1] != '(':
                result.append(' ')

        else:
            # Handle any other characters
            result.append(char)

    # Join results list and strip out any spaces in the beginning and end of the document
    formatted_sexp = ''.join(result).strip()

    # Strip out any extra space on the right hand side of each line
    formatted_sexp = '\n'.join(line.rstrip() for line in formatted_sexp.split('\n')) + '\n'

    # Formatting of s-expression completed
    return formatted_sexp

# Argument Parsing
parser = argparse.ArgumentParser(description="Prettify S-expression files")
parser.add_argument("src", type=Path, help="Source file path")
parser.add_argument("--dst", type=Path, help="Destination file path")
args = parser.parse_args()

src_file = args.src.resolve()
dst_file = args.dst.resolve() if args.dst else None

# Open source file
with open(src_file, "r") as file:
    sexp_data = file.read()

# Format the S-expression
pretty_sexpr = prettify_sexpr_minimal(sexp_data)

# Save or output the result
if dst_file:
    with open(dst_file, "w") as file:
        file.write(pretty_sexpr)
else:
    print(pretty_sexpr)

File: sexp_formatter_minimal.c
/// usr/bin/env ccache gcc -Wall -Wextra -Werror -O3 -std=gnu17 "$0" -o /tmp/sexp_formatter -lm && /tmp/sexp_formatter "$@"; exit

// KiCADv8 Style Prettify S-Expression Formatter (sexp formatter) (minimal logic version)
// By Brian Khuu, 2024
// This script reformats KiCad-like S-expressions to match a specific formatting style.
// Note: This script modifies formatting only; it does not perform linting or validation.
//       Also in addition this is a minimal version which does not support compact element handling
//       (e.g. pts element will compact xy sub elements into one line)

#include <ctype.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

/*****************************************************************************/

struct PrettifySExprState
{
    unsigned int indent;
    bool in_quote;
    bool escape_next_char;
    bool singular_element;
    bool space_pending;
    char c_out_prev;
};

void prettify_sexpr_minimal(struct PrettifySExprState *state, char c, void (*output_func)(char, void *), void *context_putc)
{
    // Parse quoted string
    if (state->in_quote || c == '"')
    {
        // Handle quoted strings
        if (state->space_pending)
        {
            // Add space before this quoted string
            output_func(' ', context_putc);
            state->space_pending = false;
        }

        if (state->escape_next_char)
        {
            // Escaped Char
            state->escape_next_char = false;
        }
        else if (c == '\\')
        {
            // Escape Next Char
            state->escape_next_char = true;
        }
        else if (c == '"')
        {
            // End of quoted string mode
            state->in_quote = !state->in_quote;
        }

        output_func(c, context_putc);
        state->c_out_prev = c;
        return;
    }

    // Parse space
    if (isspace(c))
    {
        // Handle spaces and newlines
        state->space_pending = true;
    }

    // Parse Opening parentheses
    if (c == '(')
    {
        // Handle opening parentheses
        state->space_pending = false;

        // Indented newline for this new parentheses unless at root
        if (state->indent > 0)
        {
            output_func('\n', context_putc);
            for (unsigned int j = 0; j < state->indent; ++j)
            {
                output_func('\t', context_putc);
            }
        }

        state->singular_element = true;
        state->indent++;

        output_func('(', context_putc);
        state->c_out_prev = '(';
        return;
    }

    // Parse Closing Brace
    if (c == ')')
    {
        // Handle closing parentheses
        state->space_pending = false;

        if (state->indent > 0)
        {
            state->indent--;
        }

        if (state->singular_element)
        {
            // End of singular element
            output_func(')', context_putc);
            state->singular_element = false;
        }
        else
        {
            // End of a parent element
            output_func('\n', context_putc);
            for (unsigned int j = 0; j < state->indent; ++j)
            {
                output_func('\t', context_putc);
            }
            output_func(')', context_putc);
        }

        // Add a newline if it's the end of the document
        if (state->indent <= 0)
        {
            output_func('\n', context_putc);
        }

        state->c_out_prev = ')';
        return;
    }

    // Parse Characters
    if (c != '\0')
    {
        // Handle other characters
        if (state->c_out_prev == ')')
        {
            // Indented newline for this character
            output_func('\n', context_putc);
            for (unsigned int j = 0; j < state->indent; ++j)
            {
                output_func('\t', context_putc);
            }
            state->space_pending = false;
        }
        else if (state->space_pending)
        {
            // Add space before this character
            output_func(' ', context_putc);
            state->space_pending = false;
        }

        output_func(c, context_putc);
        state->c_out_prev = c;
        return;
    }
}

/*****************************************************************************/

void putc_handler(char c, void *context_putc) { fputc(c, (FILE *)context_putc); }

// Main function
int main(int argc, char **argv)
{
    const char *prog_name = argv[0];

    if (argc == 1 || (argc == 2 && (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0)))
    {
        printf("S-Expression Formatter Minimal (Brian Khuu 2024)\n");
        printf("\n");
        printf("Usage:\n");
        printf("  %s -     -      Standard Input To Standard Output\n", prog_name);
        printf("  %s -     [dst]  Standard Input To File Output\n", prog_name);
        printf("  %s [src] -      File Input To Standard Output\n", prog_name);
        printf("  %s [src]        File Input To Standard Output\n", prog_name);
        printf("  %s [src] [dst]  File Input To File Output\n", prog_name);
        printf("\n");
        printf("Options:\n");
        printf("  -h --help       Show Help Message\n");
        return EXIT_SUCCESS;
    }

    const char *src_path = argc >= 2 ? argv[1] : NULL;
    const char *dst_path = argc >= 3 ? argv[2] : NULL;

    // Get File Descriptor
    FILE *src_file = stdin;
    if (src_path && strcmp(src_path, "-") != 0)
    {
        src_file = fopen(src_path, "r");
        if (!src_file)
        {
            perror("Error opening source file");
            return EXIT_FAILURE;
        }
    }

    // Open the destination file else default to standard output
    FILE *dst_file = stdout;
    if (dst_path && strcmp(dst_path, "-") != 0)
    {
        dst_file = fopen(dst_path, "wb");
        if (!dst_file)
        {
            perror("Error opening destination file");
            fclose(src_file);
            return EXIT_FAILURE;
        }
    }

    // Prettify the content
    char src_char;
    struct PrettifySExprState state = {0};
    while ((src_char = fgetc(src_file)) != EOF)
    {
        prettify_sexpr_minimal(&state, src_char, &putc_handler, dst_file);
    }

    // Wrapup and Cleanup
    fclose(src_file);
    if (dst_file != stdout)
    {
        fclose(dst_file);
    }

    return EXIT_SUCCESS;
}
